services/brig/src/Brig/API/Public.hs

```diff
@@ -1158,102 +1178,144 @@ getHandleInfo self handle = do
 
-changeHandle :: UserId -> ConnId -> HandleUpdate -> Handler ()
-changeHandle u conn (HandleUpdate h) = do
+changeHandle :: UserId -> ConnId -> Public.HandleUpdate -> Handler ()
+changeHandle u conn (Public.HandleUpdate h) = do
   handle <- validateHandle h
   API.changeHandle u conn handle !>> changeHandleError
 
-beginPasswordResetH :: JSON ::: JsonRequest NewPasswordReset -> Handler Response
+beginPasswordResetH :: JSON ::: JsonRequest Public.NewPasswordReset -> Handler Response
 beginPasswordResetH (_ ::: req) = do
   setStatus status201 empty <$ (beginPasswordReset =<< parseJsonBody req)
 
-beginPasswordReset :: NewPasswordReset -> Handler ()
-beginPasswordReset (NewPasswordReset target) = do
+beginPasswordReset :: Public.NewPasswordReset -> Handler ()
+beginPasswordReset (Public.NewPasswordReset target) = do
   checkWhitelist target
   (u, pair) <- API.beginPasswordReset target !>> pwResetError
   loc <- lift $ API.lookupLocale u
   lift $ case target of
     Left email -> sendPasswordResetMail email pair loc
     Right phone -> sendPasswordResetSms phone pair loc
 
-completePasswordResetH :: JSON ::: JsonRequest CompletePasswordReset -> Handler Response
+completePasswordResetH :: JSON ::: JsonRequest Public.CompletePasswordReset -> Handler Response
 completePasswordResetH (_ ::: req) = do
-  CompletePasswordReset {..} <- parseJsonBody req
+  Public.CompletePasswordReset {..} <- parseJsonBody req
   API.completePasswordReset cpwrIdent cpwrCode cpwrPassword !>> pwResetError
   return empty
 
-sendActivationCodeH :: JsonRequest SendActivationCode -> Handler Response
+sendActivationCodeH :: JsonRequest Public.SendActivationCode -> Handler Response
 sendActivationCodeH req = do
   empty <$ (sendActivationCode =<< parseJsonBody req)
 
 -- docs/reference/user/activation.md {#RefActivationRequest}
-sendActivationCode :: SendActivationCode -> Handler ()
-sendActivationCode SendActivationCode {..} = do
+sendActivationCode :: Public.SendActivationCode -> Handler ()
+sendActivationCode Public.SendActivationCode {..} = do
   checkWhitelist saUserKey
   API.sendActivationCode saUserKey saLocale saCall !>> sendActCodeError
 
-changeSelfEmailH :: UserId ::: ConnId ::: JsonRequest EmailUpdate -> Handler Response
-changeSelfEmailH (u ::: _ ::: req) = changeEmail u req True
+changeSelfEmailH :: UserId ::: ConnId ::: JsonRequest Public.EmailUpdate -> Handler Response
+changeSelfEmailH (u ::: _ ::: req) = do
+  email <- Public.euEmail <$> parseJsonBody req
+  changeSelfEmail u email >>= \case
+    ChangeEmailResponseIdempotent -> pure (setStatus status204 empty)
+    ChangeEmailResponseNeedsActivation -> pure (setStatus status202 empty)
 
-createConnectionH :: JSON ::: UserId ::: ConnId ::: JsonRequest ConnectionRequest -> Handler Response
+data ChangeEmailResponse
+  = ChangeEmailResponseIdempotent
+  | ChangeEmailResponseNeedsActivation
+
+changeSelfEmail :: UserId -> Public.Email -> Handler ChangeEmailResponse
+changeSelfEmail u email = do
```

Comment:
ah, this is the refactoring, not the internal one.  anyway, good change.  i'll rebase onto this.